### ISSUE TITLE
time control 5 seconds ?

### ISSUE THREAD

**Issue by tissatussa** [RedBedHed/Homura-Thesis-Version#4]
i let Homura play some games in CuteChess against other engines .. indeed Homura takes 5 seconds per move, it's even exactly "5.0s" each move, according to the CuteChess move listing .. Homura performs rather well, i guess its rating is about 2500 (as you're also assuming in the README), but i think this type of time management is poor : when you'd implement a "dynamic" time management, like most strong engines have, Homura could play a lot stronger !? Also those 5 seconds permit Homura to play a decent Bullet game (1 minute per player) : it will lose on time after move 12 (60/5), so it can never compete in any Bullet tournament ..

i also wonder how Homura decides `best move` after exactly 5 seconds.

i do not want to state Homura is a simple engine, but its time management is .. i think in general many really simple chess engines lack proper time management, they just want to finish search & calculations at the current depth, before deciding to go one ply deeper - if remaining time allows it .. but sometimes this (too) simple approach makes such engine lose on time : it can't afford to calculate a next depth, it will take too long, and in most cases will lead to the same `best move` as last depth calculations pointed out .. however, when a search is ended like that, the given `best move` can be considered 'accurate' (upto the last depth) because a full move list was done, not only part of it, when time's up -- you may correct me on all of this, if i'm wrong ..

letting engines play games in CuteChess GUI, i can see their progress & info & pv-lines at each depth, and sometimes it seems the engine starts calculating a next depth, but for a short time, because time's up, and then it comes up with a `best move` which was never considered before .. this is suspicious to me : was it just the `best move` until then ? (at the concerning depth) .. i think, when time's up in the middle of a depth calculation, it would be better to give `best move` of the finished previous depth ..

i read some explanations and implementations of the MultiPV (mpv) function .. i also tested many engines with this MPV "top-N `best move`s" function, setting a typical (puzzle) FEN, like the "Test Suites", and see if the engine finds the (known) solution move, and how fast, and how this move made it to the N-1 in the top-N list : often we see for many depths it was low on the top-N list or even not considered at all, and then suddenly the solution move appears to be `best move` .. without MPV the solution move might not be found, even upto higher depths .. NOTE: when playing in CuteChess, MPV is "always set 1 for best play", as the recommendation goes .. MPV will slow down the whole evaluation process and thus wil give a poor result when using limited time. 

Homura does not have MPV, and i do not promote to implement it (although MPV can be nice during development), but i mentioned this function because maybe your code manages a top-N list somehow while calculating, and picks the latest N-1 as `best move` after 5.0 seconds ?! How does it work 'internally' ? Are your `best move`s always 'accurate', as i described it ? (Or) does Homura give the `best move` of the previous completed depth ?

[ now i wrote all this, it might seem somewhat confusing .. sorry for that, hope you get my picture ]

**Comment by RedBedHed** [comment]
Hi @tissatussa,

I appreciate these tips. You make some very good points! I am hoping to add better time control logic to this repo very soon. Homura performs iterative deepening in a non-traditional way. It does many depth-limited rollouts until it can prove the value of the root at the current depth (V- >= V+). It then saves the best move, increments the depth, and starts again. If it cannot prove the value of the root at the current depth before time is up, it uses the best move from the previous depth. 

It is a bit of a strange algorithm, but it relies on much of the same theory as the classical search. Its tree policy (not always leftmost) is the main difference.

**Comment by RedBedHed** [comment]
I agree that the engine is simple. It is simple because it was written as an undergraduate research project. Although I wanted to engine to be more of a hobby project, I couldn't quite treat it that way as a student, and I ran out of time to explore other aspects of engine programming. Most of my effort was directed towards the search algorithm and trying out new ideas there. 

However, I recently graduated and upgraded my hardware. It will probably be at least a year before I can return to school, so I plan to spend a lot of free time contributing additional functionality and strength to Homura. :)

**Comment by tissatussa** [comment]
a simple function for time management, which seems good enough and is used by a lot of chess engine programmers, is the following : for each move, assume the whole game will take (only) 30 more moves until the end, and divide the remaining time by this number to get a raw but decent indication how much time should be spent at the current move .. so, when playing a 3+2 blitz game, at start the engine will use around 8 sec per move (180/30 +2) .. when the last minute arrives, this will be 4 sec (60/30 +2) .. i guess this number 30 is found to be a good value in practice, i don't know ..

i propose not to spend too much of your (coding) time on this issue, just start using the above function, i'm almost sure this will instantly improve Homura greatly.

**Comment by tissatussa** [comment]
to compete in tournaments, the format "go wtime 182000 btime 182000 winc 2000 binc 2000" (eg. for a 3+2 blitz game) will often be used .. i like the bonus time, also in my own play : when a position "is winning", it can be won .. i guess playing a (blitz) game without bonus needs a different time management strategy .. a recent version of the UCI Chess Engine Protocol i found here : https://github.com/fsmosca/UCIChessEngineProtocol (August 2021).
